cub/benchmarks/bench/segmented_reduce/base.cuh

```diff
@@ -0,0 +1,154 @@
+template <typename AccumT>
+struct policy_hub_t
+{
+  struct policy_t : cub::ChainedPolicy<300, policy_t, policy_t>
+  {
+    static constexpr int threads_per_block  = TUNE_THREADS_PER_BLOCK;
+    static constexpr int items_per_thread   = TUNE_ITEMS_PER_THREAD;
+    static constexpr int items_per_vec_load = TUNE_ITEMS_PER_VEC_LOAD;
+
+    static constexpr int small_threads_per_warp = TUNE_S_THREADS_PER_WARP;
+    static constexpr int small_items_per_thread = TUNE_S_ITEMS_PER_THREAD;
+
+    static constexpr int medium_threads_per_warp = TUNE_M_THREADS_PER_WARP;
+    static constexpr int medium_items_per_thread = TUNE_M_ITEMS_PER_THREAD;
+
+    using SmallReducePolicy =
+      cub::AgentWarpReducePolicy<threads_per_block,
+                                 medium_threads_per_warp,
+                                 medium_items_per_thread,
+                                 AccumT,
+                                 items_per_vec_load,
+                                 cub::LOAD_LDG>;
+    using MediumReducePolicy =
+      cub::AgentWarpReducePolicy<threads_per_block,
+                                 small_threads_per_warp,
+                                 small_items_per_thread,
+                                 AccumT,
+                                 items_per_vec_load,
+                                 cub::LOAD_LDG>;
+    using ReducePolicy =
+      cub::AgentReducePolicy<threads_per_block,
+                             items_per_thread,
+                             AccumT,
+                             items_per_vec_load,
+                             cub::BLOCK_REDUCE_WARP_REDUCTIONS,
+                             cub::LOAD_LDG>;
+  };
+
+  using MaxPolicy = policy_t;
+};
+#endif // !TUNE_BASE
+
+template <typename T, typename OffsetT>
+void fixed_size_segmented_reduce(nvbench::state& state, nvbench::type_list<T, OffsetT>)
+{
+  using accum_t        = T;
+  using input_it_t     = const T*;
+  using output_it_t    = T*;
+  using offset_t       = cub::detail::choose_offset_t<OffsetT>;
+  using segment_size_t = offset_t;
+  using output_t       = T;
+  using init_t         = T;
+
+#if !TUNE_BASE
+  using policy_t   = policy_hub_t<accum_t>;
+  using dispatch_t = cub::detail::reduce::
+    DispatchFixedSizeSegmentedReduce<input_it_t, output_it_t, offset_t, segment_size_t, op_t, init_t, accum_t, policy_t>;
+#else // TUNE_BASE
+  using dispatch_t = cub::detail::reduce::
+    DispatchFixedSizeSegmentedReduce<input_it_t, output_it_t, offset_t, segment_size_t, op_t, init_t, accum_t>;
+#endif // TUNE_BASE
+
+  // Retrieve axis parameters
+  const size_t num_segments = static_cast<size_t>(state.get_int64("NumSegments{io}"));
+  const size_t segment_size = static_cast<size_t>(state.get_int64("SegmentSize{io}"));
+  const size_t elements     = num_segments * segment_size;
+
+  const size_t max_elements = std::pow(2, 30);
+  if (elements > max_elements)
+  {
+    state.skip("Input size too large for current configuration");
+    return;
+  }
```

Comment:
question: would it help in any way if we fixed the number of elements to, say, 2^28 and only change the number of segments in this benchmark?

Reply:
We can do that, but it will reduce the number of data points signficantly and we get performance numbers only on one input size (`2^28 * sizeof(T)`) in total.

Reply:
If we parametrize the test by `NumElements{io}`, and compute problem parameters like so:

```cpp
const size_t input_nelems = static_cast<size_t>(state.get_int64("NumElements{io}"));
const size_t segment_size = static_cast<size_t>(state.get_int64("SegmentSize{io}"));
const size_t num_segments = std::max<std::size_t>(1, (input_num_elements / segment_size));
const size_t elements = num_segments * segment_size;
```

we will not need to skip any configuration points.